### PR TITLE
Add HTTP CONNECT proxy support for LimaCharlie connections

### PR DESCRIPTION
**Pull request by maximelb**
## Summary
- Add support for connecting to LimaCharlie cloud through HTTP CONNECT proxies
- Add ProxyURL, ProxyUsername, and ProxyPassword fields to ClientOptions
- Implement automatic retry logic (3 retries) for proxy connections

## Implementation Details
- Created `createProxyDialer()` method that establishes HTTP CONNECT tunnel
- Supports HTTP Basic authentication for proxy servers
- Handles TLS/SSL negotiation after tunnel establishment
- Only affects WebSocket connections to LimaCharlie cloud

## Configuration
Proxy settings can be configured via:
- YAML/JSON configuration files
- Command-line parameters
- Environment variables

Example:
```yaml
client_options:
  proxy_url: "http://proxy.example.com:8080"
  proxy_username: "user"  # optional
  proxy_password: "pass"  # optional
```

## Test plan
- [ ] Test direct connection (no proxy)
- [ ] Test connection through proxy without authentication
- [ ] Test connection through proxy with authentication
- [ ] Test retry logic on proxy connection failures
- [ ] Verify TLS/SSL works correctly through proxy

🤖 Generated with Claude Code

### CODE REVIEW

**Comment by maximelb** [comment]
## Code Review

Overall, this is a well-implemented feature that adds HTTP CONNECT proxy support for LimaCharlie connections. The implementation is clean and follows good practices. Here are my findings:

### ✅ Strengths

1. **Well-structured implementation**: The `createProxyDialer()` method is cleanly separated and handles the proxy logic appropriately
2. **Security considerations**: Proper support for proxy authentication using HTTP Basic Auth
3. **Error handling**: Good error messages and proper cleanup of connections on failure
4. **Retry logic**: Smart addition of retry mechanism (3 retries) specifically for proxy connections
5. **TLS handling**: Correctly handles TLS/SSL negotiation after establishing the proxy tunnel

### 🔍 Suggestions for Improvement

1. **Test Coverage** ⚠️
   - The PR lacks tests for the new proxy functionality
   - Consider adding unit tests for `createProxyDialer()`
   - Integration tests for proxy scenarios would be valuable

2. **Documentation**
   - Consider updating the README.md to document the new proxy configuration options
   - Add usage examples for proxy configuration

3. **Minor Code Improvements**

   a. **Resource cleanup**: Consider using `defer` for connection cleanup in error cases:
   ```go
   proxyConn, err := net.DialTimeout("tcp", proxyURL.Host, 10*time.Second)
   if err \!= nil {
       return nil, fmt.Errorf("failed to connect to proxy: %v", err)
   }
   defer func() {
       if err \!= nil {
           proxyConn.Close()
       }
   }()
   ```

   b. **Context propagation**: The `NetDialContext` receives a context but doesn't use it for the TCP dial:
   ```go
   // Current:
   proxyConn, err := net.DialTimeout("tcp", proxyURL.Host, 10*time.Second)
   
   // Consider:
   var d net.Dialer
   d.Timeout = 10 * time.Second
   proxyConn, err := d.DialContext(ctx, "tcp", proxyURL.Host)
   ```

   c. **Validation**: Consider adding validation for the proxy URL in `ClientOptions.Validate()`:
   ```go
   if o.ProxyURL \!= "" {
       if _, err := url.Parse(o.ProxyURL); err \!= nil {
           return fmt.Errorf("invalid proxy URL: %v", err)
       }
   }
   ```

4. **Environment Variable Support**
   - The current implementation doesn't respect standard proxy environment variables (HTTP_PROXY, HTTPS_PROXY, NO_PROXY)
   - Consider whether this is intentional or if environment variable support should be added

5. **Configuration Clarity**
   - The `websocket.Dialer` is created with `Proxy: http.ProxyFromEnvironment` but then overrides with custom `NetDialContext`
   - This might be confusing - consider setting `Proxy: nil` explicitly

### 📋 Testing Checklist

Before merging, please ensure:
- [ ] Unit tests for `createProxyDialer()` are added
- [ ] Manual testing with actual proxy server is completed
- [ ] Error scenarios (proxy auth failure, connection timeout) are tested
- [ ] Documentation is updated

### Verdict

The implementation is solid and ready for use. The main gap is test coverage, which should be addressed before merging to ensure long-term maintainability. Great work on the retry logic and proper handling of the HTTP CONNECT protocol\! 👍

**Comment by maximelb** [comment]
## Code Review Feedback Addressed ✅

Thank you for the thorough code review\! I've addressed all the suggestions (except environment variable support as requested):

### 🔧 Code Improvements Implemented

1. **Resource cleanup with defer** ✅
   - Added proper defer cleanup for proxy connections on error
   - Ensures connections are always closed if an error occurs during setup

2. **Context propagation** ✅
   - Changed from `net.DialTimeout` to `d.DialContext(ctx, "tcp", proxyURL.Host)`
   - Now properly respects the context passed to `NetDialContext`

3. **Proxy URL validation** ✅
   - Added validation in `ClientOptions.Validate()`
   - Validates proxy URL format
   - Ensures both username and password are provided together for authentication

4. **Explicit Proxy setting** ✅
   - Set `Proxy: nil` in the dialer configuration
   - Makes it clear we handle proxy logic via `NetDialContext`

### 🧪 Test Coverage Added

Created comprehensive test suite in `client_proxy_test.go`:
- `TestClientOptions_Validate_ProxyConfig` - Tests proxy configuration validation
- `TestCreateProxyDialer` - Tests dialer creation with various proxy configs
- `TestProxyDialerConnection` - Tests successful proxy connections
- `TestProxyDialerAuthFailure` - Tests authentication failure scenarios
- `TestInvalidProxyURL` - Tests invalid proxy URL handling

All tests are passing ✅

### 📚 Documentation Updated

Updated `README.md` with:
- Added proxy support to features list
- Created dedicated "Proxy Support" section
- Included configuration examples
- Documented validation requirements

### 🎯 Testing Checklist Progress

- [x] Unit tests for `createProxyDialer()` are added
- [ ] Manual testing with actual proxy server (ready for testing)
- [x] Error scenarios (proxy auth failure, invalid URLs) are tested
- [x] Documentation is updated

The implementation is now more robust with better error handling, proper resource management, and comprehensive test coverage. Ready for another review\!

File: README.md [hunk 1]
```diff
@@ -1,6 +1,45 @@
...
+```go
+options := uspclient.ClientOptions{
+    Identity: uspclient.Identity{
+        Oid:             "your-org-id",
+        InstallationKey: "your-key",
+    },
+    Platform: "your-platform",
+    
+    // Proxy configuration
+    ProxyURL:      "http://proxy.example.com:8080",
```
Comment (tomaz-lc) [inline README.md:26]: Minor thing, would it make sense to use a nested`proxy` top level key (e.g. `proxy.{URL,username,password}`).

This way it would be more consistent with `identity` which also uses nested options.

Comment (tomaz-lc) [inline README.md:26]: We don't support credentials in the URL itself, just specified explicitly via the username + password option, right?

If so, it would not be a bad idea to throw in case credentials are specified in the URL itself (if a parsed URL object contains `User` field).

File: client.go [hunk 2]
```diff
@@ -202,6 +225,84 @@ func (c *Client) Close() ([]*protocol.DataMessage, error) {
...
+	}
+
+	proxyURL, err := url.Parse(c.options.ProxyURL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid proxy URL: %v", err)
+	}
+
+	dialer := &websocket.Dialer{
+		Proxy:            nil, // We handle proxy ourselves via NetDialContext
+		HandshakeTimeout: 45 * time.Second,
```
Comment (tomaz-lc) [inline client.go:242]: Should we make this configurable via `Extra` or similar client options (or even under top level `proxy` option, e.g. `proxy.handshake_timeout`).

File: client.go [hunk 3]
```diff
@@ -202,6 +225,84 @@ func (c *Client) Close() ([]*protocol.DataMessage, error) {
...
+		NetDialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			// Connect to proxy with context
+			var d net.Dialer
+			d.Timeout = 10 * time.Second
```
Comment (tomaz-lc) [inline client.go:246]: Same here, probably not a bad idea to make it configurable / overridable via extra / advanced client config options.

Comment (tomaz-lc) [reply client.go:246]: And it would also be a good idea to set a timeout for the actual reads and writes aka `proxyConn.SetDeadline`

**Comment by maximelb** [comment]
@tomaz-lc I've pushed another commit that removes the backward compatibility code entirely, keeping the implementation clean and simple since this feature isn't in use yet.

The latest changes:
- Removed deprecated `ProxyURL`, `ProxyUsername`, and `ProxyPassword` fields
- Simplified the code to only use the new `Proxy` nested configuration
- Cleaned up tests and documentation

This makes the implementation much cleaner while still addressing all your original feedback. The proxy configuration now follows a single, consistent pattern.

File: client_proxy_test.go [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+	defer conn.Close()
+
+	reader := bufio.NewReader(conn)
+	request, err := http.ReadRequest(reader)
+	if err != nil {
+		p.t.Logf("Failed to read request: %v", err)
+		return
+	}
+
+	if request.Method != "CONNECT" {
```
Comment (tomaz-lc) [inline client_proxy_test.go:63]: (not a blocker) I think this is good so we can test various normal and edge cases, but it would also be nice to also have some end to end tests against some actual proxy software (e.g. squid).

**Review by tomaz-lc — APPROVED** [review]
